### PR TITLE
Add support to DELETE custom search parameters

### PR DESCRIPTION
**Pull request by feordin**
## Description
Adds the support to the DeleteResourceHandler to remove custom search parameter from the SearchParameterDefinitionManager collections.

## Related issues
Addresses [issue [AB#76347](https://microsofthealth.visualstudio.com/f8da5110-49b1-4e9f-9022-2f58b6124ff9/_workitems/edit/76347)].

## Testing
Unit test added. E2E Test updated with DELETE.  Manual testing performed.

## FHIR Team Checklist
- [X] **Update the title** of the PR to be succinct and less than 50 characters
- [x] **Add a milestone** to the PR for the sprint that it is merged (i.e. add S47)
- [x] Tag the PR with the type of update: **Bug**, **Dependencies**, **Enhancement**, or **New-Feature**
- [x] Tag the PR with **Azure API for FHIR** if this will release to the managed service
- Review [squash-merge requirements](https://github.com/microsoft/fhir-server/blob/master/SquashMergeRequirements.md)

### Semver Change ([docs](https://github.com/microsoft/fhir-server/blob/master/docs/Versioning.md))
Patch|Skip|Feature|Breaking (reason)

### CODE REVIEW

File: src/Microsoft.Health.Fhir.Core/Extensions/RawResourceExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+{
+    public static class RawResourceExtensions
+    {
+        /// <summary>
+        /// Converts the RawResource to an ITypedElement
+        /// </summary>
+        /// <param name="rawResource">The RawResource to convert</param>
+        /// <param name="modelInfoProvider">The IModelInfoProvider to use when converting the RawResource</param>
+        /// <returns>An ITypedElement of the RawResource</returns>
+        public static ITypedElement ToITypedElement(this RawResource rawResource, IModelInfoProvider modelInfoProvider)
```
Comment (Ivanidzo4ka) [inline src/Microsoft.Health.Fhir.Core/Extensions/RawResourceExtensions.cs:26]: >ToITypedElement [](start = 36, length = 15)

This looks quite similar to `ReadEmbeddedSearchParameters` in `SearchParameterDefinitionBuilder`
Is it worth to consider using this method in that part of code?

File: src/Microsoft.Health.Fhir.Core/Features/Definition/SearchParameterDefinitionManager.cs [hunk 2]
```diff
@@ -151,5 +152,27 @@ private void CalculateSearchParameterHash()
...
+                throw new Exception(string.Format(Resources.CustomSearchParameterNotfound, searchParamWrapper.Url));
+            }
+
+            var searchParameterInfo = TypeLookup[searchParamWrapper.Type][searchParamWrapper.Name];
+            var allResourceTypes = searchParameterInfo.TargetResourceTypes.Union(searchParameterInfo.BaseResourceTypes);
+            foreach (var resourceType in allResourceTypes)
+            {
+                if (TypeLookup.ContainsKey(resourceType))
+                {
+                    TypeLookup[resourceType].Remove(searchParameterInfo.Name);
```
Comment (Ivanidzo4ka) [inline src/Microsoft.Health.Fhir.Core/Features/Definition/SearchParameterDefinitionManager.cs:171]: >.Remove [](start = 44, length = 7)

How we protect ourselves from concurrency issues?
From what I can see in code, that `IDictionary<string, SearchParameterInfo>` is just a regular Dictionary, and it's ok if you instantiate it once during startup and just do lookup for it, but now we adding and removing elements from it, and I'm not sure is it locked for one thread in upstream or just not locked at at all.

File: src/Microsoft.Health.Fhir.Core/Resources.resx [hunk 3]
```diff
@@ -515,4 +515,14 @@
...
   </data>
-</root>
+  <data name="CustomSearchDeleteError" xml:space="preserve">
+    <value>An error occurred deleting the custom search parameter.  The issue must be resolved and the delete request resubmitted to remove the parameter.</value>
+  </data>
+  <data name="CustomSearchParameterNotfound" xml:space="preserve">
+    <value>Custom Search parameter with Uri: {0} was not found.</value>
+    <comment>{0} the uri identifying the custom search parameter.</comment>
+  </data>
+  <data name="NoResourcesNeedbeReindexed" xml:space="preserve">
```
Comment (Ivanidzo4ka) [inline src/Microsoft.Health.Fhir.Core/Resources.resx:525]: >be [](start = 29, length = 2)

super nit: typo?

File: src/Microsoft.Health.Fhir.Core/Features/Definition/BundleWrappers/SearchParameterWrapper.cs [hunk 4]
```diff
@@ -14,7 +14,7 @@
 
 namespace Microsoft.Health.Fhir.Core.Features.Definition.BundleWrappers
 {
-    internal class SearchParameterWrapper
+    public class SearchParameterWrapper
```
Comment (Ivanidzo4ka) [inline src/Microsoft.Health.Fhir.Core/Features/Definition/BundleWrappers/SearchParameterWrapper.cs:17]: >public [](start = 4, length = 6)

Is this change necessary?

File: src/Microsoft.Health.Fhir.Core/Features/Search/Parameters/DeleteSearchParameterBehavior.cs [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+            {
+                searchParamResource = await _fhirDataStore.GetAsync(deleteRequest.ResourceKey, cancellationToken);
+            }
+
+            var response = await next();
+
+            if (searchParamResource != null)
+            {
+                // Once the SearchParameter resource is removed from the data store, we can update the in
+                // memory SearchParameterDefinitionManager, and persist the deleted status to the data store
```
Comment (Ivanidzo4ka) [inline src/Microsoft.Health.Fhir.Core/Features/Search/Parameters/DeleteSearchParameterBehavior.cs:47]: >persist the deleted status to the data store [](start = 64, length = 44)

Didn't we remove it from data store already?
This part a bit confusing.

File: src/Microsoft.Health.Fhir.Core/Features/Search/Parameters/DeleteSearchParameterBehavior.cs [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+        {
+            var deleteRequest = request as DeleteResourceRequest;
+            ResourceWrapper searchParamResource = null;
+
+            if (deleteRequest.ResourceKey.ResourceType.Equals(KnownResourceTypes.SearchParameter, StringComparison.Ordinal))
+            {
+                searchParamResource = await _fhirDataStore.GetAsync(deleteRequest.ResourceKey, cancellationToken);
+            }
+
+            var response = await next();
```
Comment (Ivanidzo4ka) [inline src/Microsoft.Health.Fhir.Core/Features/Search/Parameters/DeleteSearchParameterBehavior.cs:42]: >next [](start = 33, length = 4)

nit:
Part of me want to argue about order of things.
So if correctly understand there is two parts, storing search parameter in data store, and storing search parameter in memory of the server.
By having search parameter in memory, we apply it to all incoming data, and i think we should purge it from memory first and then try to remove it from server. If someone try to delete search parameter i would assume they don't want it to be applied anymore, so it sound a bit unnatural for me to remove it from memory as second step.
Also if I look into adding search parameter order is : add to data store, then add to memory.
Considering what delete operation is opposite of adding,  it doesn't feel like it should follow same order, and should be opposite.

File: src/Microsoft.Health.Fhir.Core/Features/Search/Registry/SearchParameterStatusManager.cs [hunk 7]
```diff
@@ -126,5 +126,11 @@ public async Task AddSearchParameterStatusAsync(IReadOnlyCollection<string> sear
             // they will be fully enabled
             await UpdateSearchParameterStatusAsync(searchParamUris, SearchParameterStatus.Supported);
         }
+
+        internal async Task DeleteSearchParameterStatusAsync(string url)
```
Comment (Ivanidzo4ka) [inline src/Microsoft.Health.Fhir.Core/Features/Search/Registry/SearchParameterStatusManager.cs:130]: >internal [](start = 8, length = 8)

nit: it's a bit strange to see Delete as `internal` and Add as `public`.
Can they be synced?

File: src/Microsoft.Health.Fhir.Core/Features/Definition/SearchParameterDefinitionManager.cs [hunk 8]
```diff
@@ -142,14 +143,33 @@ public void AddNewSearchParameters(IReadOnlyCollection<ITypedElement> searchPara
...
+
+            if (!UrlLookup.TryRemove(new Uri(searchParamWrapper.Url), out searchParameterInfo))
+            {
+                throw new Exception(string.Format(Resources.CustomSearchParameterNotfound, searchParamWrapper.Url));
+            }
+
+            var allResourceTypes = searchParameterInfo.TargetResourceTypes.Union(searchParameterInfo.BaseResourceTypes);
+            foreach (var resourceType in allResourceTypes)
+            {
+                TypeLookup[resourceType].TryRemove(searchParameterInfo.Name, out var removedParam);
```
Comment (Ivanidzo4ka) [inline src/Microsoft.Health.Fhir.Core/Features/Definition/SearchParameterDefinitionManager.cs:169]: >TryRemove [](start = 41, length = 9)

if all `TryRemove` return false, it's probably doesn't make sense to do `CalculateSearchParameterHash`

File: test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/Search/CustomSearchParamTests.cs [hunk 9]
```diff
@@ -44,7 +44,25 @@ public async Task GivenANewSearchParam_WhenReindexingComplete_ThenResourcesSearc
...
+                // if the SearchParameter exists, we should delete it and recreate it
+                var searchParamBundle = await Client.SearchAsync(ResourceType.SearchParameter, $"url={searchParam.Url}");
+                if (searchParamBundle.Resource?.Entry[0] != null && searchParamBundle.Resource?.Entry[0].Resource.ResourceType == ResourceType.SearchParameter)
+                {
+                    await DeleteSearchParameterAndVerify(searchParamBundle.Resource?.Entry[0].Resource as SearchParameter);
+                    searchParamPosted = await Client.CreateAsync(searchParam);
+                }
+                else
+                {
+                    throw ex;
```
Comment (Ivanidzo4ka) [inline test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/Search/CustomSearchParamTests.cs:63]: >throw ex [](start = 20, length = 8)

nit: I think this is antipattern.
`ex` is not used anywhere and just got rethrown which would lead to lost of original stack. Not that important in tests, but i think it's nice thing to avoid overall.

here some details:
[https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2200](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2200)

File: src/Microsoft.Health.Fhir.Core/Features/Definition/SearchParameterDefinitionManager.cs (hunk 8 above)
Comment (feordin) [reply src/Microsoft.Health.Fhir.Core/Features/Definition/SearchParameterDefinitionManager.cs:169]: This is true, but if we passed the first attempt to remove a SearchParameter from the UrlLookup, we should have at least one entry in the TypeLookup to remove.

File: src/Microsoft.Health.Fhir.Core/Features/Search/Parameters/DeleteSearchParameterBehavior.cs (hunk 6 above)
Comment (feordin) [reply src/Microsoft.Health.Fhir.Core/Features/Search/Parameters/DeleteSearchParameterBehavior.cs:42]: I had the same thought about order and considered the alternate approach.  Currently the call to DeleteSearchParameterStatusAsync uses the metadata stored in the SearchParameterDefinitionManager to then go and remove the status from the data store.  So, because of dependencies, it does need to happen in this order.

Comment (Ivanidzo4ka) [reply src/Microsoft.Health.Fhir.Core/Features/Search/Parameters/DeleteSearchParameterBehavior.cs:42]: Would it worth to add comment regarding order? I think we both agree it's a bit counterintuitive, so maybe it's worth to leave some information here.

---
In reply to: [560619433](https://github.com/microsoft/fhir-server/pull/1581#discussion_r560619433) [](ancestors = 560619433)

**Comment by Ivanidzo4ka** [comment]
>                 Assert.True(delayCount <= 9);

I think there is change to not run job if it doesn't have any matching resources, and fails here.

Probably according to new contract this job would be considered `Cancelled`
```
 await CompleteJobAsync(OperationStatus.Canceled, cancellationToken);
```
instead of `Completed`

---
Refers to: test/Microsoft.Health.Fhir.Shared.Tests.Integration/Features/Operations/Reindex/ReindexJobTests.cs:224 in f87a5ff. [](commit_id = f87a5ff33e317f75d90b5f4cbe1803f9353ca66f, deletion_comment = False)

**Review by Ivanidzo4ka — APPROVED** [review]
:shipit:

File: src/Microsoft.Health.Fhir.Core/Features/Search/Parameters/CreateSearchParameterBehavior.cs [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+using System.Threading.Tasks;
+using EnsureThat;
+using MediatR;
+using Microsoft.Health.Fhir.Core.Messages.Create;
+using Microsoft.Health.Fhir.Core.Models;
+
+namespace Microsoft.Health.Fhir.Core.Features.Search.Parameters
+{
+    public class CreateSearchParameterBehavior<TCreateResourceRequest, TUpsertResourceResponse> : IPipelineBehavior<TCreateResourceRequest, TUpsertResourceResponse>
+        where TCreateResourceRequest : CreateResourceRequest
```
Comment (Ivanidzo4ka) [inline src/Microsoft.Health.Fhir.Core/Features/Search/Parameters/CreateSearchParameterBehavior.cs:17]: >TCreateResourceRequest [](start = 14, length = 22)

Silly question, but should it work with conditional update/create?

File: src/Microsoft.Health.Fhir.Core/Extensions/RawResourceExtensions.cs [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+            }
+            catch (FormatException ex)
+            {
+                var issue = new OperationOutcomeIssue(
+                    OperationOutcomeConstants.IssueSeverity.Fatal,
+                    OperationOutcomeConstants.IssueType.Invalid,
+                    ex.Message);
+
+                throw new InvalidDefinitionException(
+                    Core.Resources.SearchParameterDefinitionContainsInvalidEntry,
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Core/Extensions/RawResourceExtensions.cs:46]: This error message doesn't seem correct here

File: src/Microsoft.Health.Fhir.Core/Features/Definition/BundleWrappers/SearchParameterWrapper.cs [hunk 12]
```diff
@@ -45,7 +45,9 @@ public SearchParameterWrapper(ITypedElement searchParameter)
 
         public string Description => _description.Value;
 
+#pragma warning disable CA1056 // Uri properties should not be strings
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Core/Features/Definition/BundleWrappers/SearchParameterWrapper.cs:48]: Disable with attribute instead

File: src/Microsoft.Health.Fhir.Core/Extensions/RawResourceExtensions.cs (hunk 1 above)
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Core/Extensions/RawResourceExtensions.cs:26]: By switching this around it could become an overload, that might make discovery easier, thoughts?

```suggestion
        public static ITypedElement ToTypedElement(this IModelInfoProvider modelInfoProvider, RawResource rawResource)
```

File: src/Microsoft.Health.Fhir.Core/Features/Search/Parameters/CreateSearchParameterBehavior.cs [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+using EnsureThat;
+using MediatR;
+using Microsoft.Health.Fhir.Core.Messages.Create;
+using Microsoft.Health.Fhir.Core.Models;
+
+namespace Microsoft.Health.Fhir.Core.Features.Search.Parameters
+{
+    public class CreateSearchParameterBehavior<TCreateResourceRequest, TUpsertResourceResponse> : IPipelineBehavior<TCreateResourceRequest, TUpsertResourceResponse>
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Core/Features/Search/Parameters/CreateSearchParameterBehavior.cs:16]: Unless the generic is needed for derived types, I think this should still work with:
```csharp
CreateSearchParameterBehavior : IPipelineBehavior<CreateResourceRequest, UpsertResourceResponse>
```

Comment (feordin) [reply src/Microsoft.Health.Fhir.Core/Features/Search/Parameters/CreateSearchParameterBehavior.cs:16]: I may want to update this for conditionalcreate, so I wanted to leave it generic for now.

File: src/Microsoft.Health.Fhir.Core/Features/Search/Parameters/CreateSearchParameterBehavior.cs (hunk 10 above)
Comment (feordin) [reply src/Microsoft.Health.Fhir.Core/Features/Search/Parameters/CreateSearchParameterBehavior.cs:17]: I hope so :), I have not net yet tried it.

File: src/Microsoft.Health.Fhir.Core/Extensions/RawResourceExtensions.cs (hunk 1 above)
Comment (feordin) [reply src/Microsoft.Health.Fhir.Core/Extensions/RawResourceExtensions.cs:26]: I agree, I have updated it as an overload on ModelInfoProvider.
